### PR TITLE
[postfilter-impl-1] Refactor scheduler preempt interface 

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What type of PR is this?**

/kind feature
/sig scheduling

**What this PR does / why we need it**:

This is the first PR of defaultpreemption plugin, which covers:

- Replace `error` with `NodeToStatusMap` in `Preempt()` signature
- Eliminate `podPreemptor` interface and expose its original functions statelessly
- Move logic in `scheduler.go#preempt()` to `generic_scheduler.go#Preempt()`

It corresponds to the first part of @ahg-g's https://github.com/kubernetes/kubernetes/pull/91426#issuecomment-641249300

**Which issue(s) this PR fixes**:

Part of #91038.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by Huang-Wei** [comment]
/assign @ahg-g 

It's ready for review.

File: pkg/scheduler/core/generic_scheduler_test.go [hunk 1]
```diff
@@ -2435,31 +2447,39 @@ func TestPreempt(t *testing.T) {
...
 				}
 				if !found {
-					t.Errorf("pod %v is not expected to be a victim.", victim.Name)
+					t.Errorf("pod %v is not expected to be a victim.", victimName)
 				}
-				// Mark the victims for deletion and record the preemptor's nominated node name.
-				now := metav1.Now()
-				victim.DeletionTimestamp = &now
 				test.pod.Status.NominatedNodeName = node
+				client.CoreV1().Pods(test.pod.Namespace).Update(context.TODO(), test.pod, metav1.UpdateOptions{})
```
Comment (ahg-g) [inline pkg/scheduler/core/generic_scheduler_test.go:2465]: why are we doing this inside the loop iterating over the victims?

File: pkg/scheduler/core/generic_scheduler_test.go [hunk 2]
```diff
@@ -2435,31 +2447,39 @@ func TestPreempt(t *testing.T) {
...
 			}
+
+			// Manually set the deleted Pods' deletionTimestamp to non-nil.
+			for _, pod := range test.pods {
+				if deletedPodNames.Has(pod.Name) {
+					now := metav1.Now()
+					pod.DeletionTimestamp = &now
+					deletedPodNames.Delete(pod.Name)
```
Comment (ahg-g) [inline pkg/scheduler/core/generic_scheduler_test.go:2473]: why not do this inside the loop above that iterates over deletedPodNames?

**Review by ahg-g — COMMENTED** [review]
Thanks!

File: pkg/scheduler/util/utils.go [hunk 3]
```diff
@@ -109,3 +115,55 @@ func GetPodAntiAffinityTerms(affinity *v1.Affinity) (terms []v1.PodAffinityTerm)
...
+// RemoveNominatedNodeName internally submit a patch request to API server
+// to set <pod.Status.NominatedNodeName> to "".
+func RemoveNominatedNodeName(cs kubernetes.Interface, pod *v1.Pod) error {
+	if len(pod.Status.NominatedNodeName) == 0 {
+		return nil
+	}
+	podCopy := pod.DeepCopy()
+	podCopy.Status.NominatedNodeName = ""
+	return PatchPod(cs, pod, podCopy)
+}
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/util/utils.go:158]: SG.

File: pkg/scheduler/util/utils.go [hunk 4]
```diff
@@ -109,3 +115,55 @@ func GetPodAntiAffinityTerms(affinity *v1.Affinity) (terms []v1.PodAffinityTerm)
...
+	if err != nil {
+		return fmt.Errorf("failed to create merge patch for pod %q/%q: %v", old.Namespace, old.Name, err)
+	}
+	_, err = cs.CoreV1().Pods(old.Namespace).Patch(context.TODO(), old.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{}, "status")
+	return err
+}
+
+// GetUpdatedPod returns the latest version of <pod> from API server.
+func GetUpdatedPod(cs kubernetes.Interface, pod *v1.Pod) (*v1.Pod, error) {
+	return cs.CoreV1().Pods(pod.Namespace).Get(context.TODO(), pod.Name, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/util/utils.go:141]: Let me add a todo.

File: pkg/scheduler/core/generic_scheduler_test.go (hunk 1 above)
Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler_test.go:2465]: As mentioned in your previous comment, we're now still fetching the pod from clientset, so the test needs to update the pod in fake clienset as well.

File: pkg/scheduler/core/generic_scheduler_test.go (hunk 2 above)
Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler_test.go:2473]: The `test.expectedPods` in the above loop is a string slice, so I cannot update the pod there.

File: pkg/scheduler/core/generic_scheduler_test.go (hunk 1 above)
Comment (ahg-g) [reply pkg/scheduler/core/generic_scheduler_test.go:2465]: sure, but why are we doing the update inside the loop iterating over ```deletedPodNames```, shouldn't the update happen once after or before the loop?

Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler_test.go:2465]: Ah, the old logic which assigns "node" to "test.pod.Status.NominatedNodeName" misleads me. Updated.

PS: technically I can make `expectedPods` as a string set to shorten the code, but I have a PR on refactoring the whole TestPreempt testcase, which'd make this more neat. So for now, I'd prefer to just make minimum change.

File: pkg/scheduler/core/generic_scheduler_test.go (hunk 2 above)
Comment (Huang-Wei) [reply pkg/scheduler/core/generic_scheduler_test.go:2473]: As mentioned in https://github.com/kubernetes/kubernetes/pull/92009#discussion_r439803037, there will be a PR refactoring whole TestPreempt test later.

**Comment by ahg-g** [comment]
Thanks, please squash.

**Comment by Huang-Wei** [comment]
/retes

@ahg-g Thanks for reviewing! Commits squashed. PTAL.

**Pull request merged**